src/compiler/checker.rs

```diff
@@ -816,7 +896,74 @@ impl TypeChecker {
     }
 }
 
-type ErrorReporter<'a> = &'a dyn FnMut(DiagnosticMessage);
+fn create_node_builder() -> NodeBuilder {
+    NodeBuilder::new()
+}
+
+pub struct NodeBuilder {}
+
+impl NodeBuilder {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    pub fn type_to_type_node(
+        &self,
+        type_: &Type,
+        tracker: Option<&dyn SymbolTracker>,
+    ) -> Option<TypeNode> {
+        self.with_context(tracker, |context| {
+            self.type_to_type_node_helper(type_, context)
+        })
+    }
+
+    fn with_context<TReturn, TCallback: FnMut(&NodeBuilderContext) -> TReturn>(
+        &self,
+        tracker: Option<&dyn SymbolTracker>,
+        mut cb: TCallback,
+    ) -> Option<TReturn> {
+        let context = NodeBuilderContext::new(tracker.unwrap());
+        let resulting_node = cb(&context);
+        Some(resulting_node)
+    }
+
+    pub fn type_to_type_node_helper(&self, type_: &Type, context: &NodeBuilderContext) -> TypeNode {
+        let synthetic_factory = get_synthetic_factory();
+        if type_.flags().intersects(TypeFlags::Number) {
+            return Into::<KeywordTypeNode>::into(
+                factory.create_keyword_type_node(&synthetic_factory, SyntaxKind::NumberKeyword),
+            )
+            .into();
+        }
+        if type_.flags().intersects(TypeFlags::BooleanLiteral) {
+            return factory
+                .create_literal_type_node(
+                    &synthetic_factory,
+                    &*Into::<Rc<Node>>::into(
```

Comment:
Here I ran into it later trying to unwrap these nodes to their `Rc<Node>` wrapper where that wrapper didn't exist since these are "artificially constructed" AST nodes (for error display of eg `'number'`) rather than parsed AST nodes. So had to make sure to wrap them before passing them along